src/pybind/mgr/dashboard/frontend/src/app/ceph/smb/smb-cluster-list/smb-cluster-list.component.ts

```diff
@@ -0,0 +1,107 @@
+import { Icons } from '~/app/shared/enum/icons.enum';
+import { CdTableAction } from '~/app/shared/models/cd-table-action';
+import { CdTableColumn } from '~/app/shared/models/cd-table-column';
+import { CdTableSelection } from '~/app/shared/models/cd-table-selection';
+import { Permission } from '~/app/shared/models/permissions';
+import { ListWithDetails } from '~/app/shared/classes/list-with-details.class';
+import { AuthStorageService } from '~/app/shared/services/auth-storage.service';
+import { SmbService } from '~/app/shared/api/smb.service';
+import { CdTableFetchDataContext } from '~/app/shared/models/cd-table-fetch-data-context';
+
+@Component({
+  selector: 'cd-smb-cluster-list',
+  templateUrl: './smb-cluster-list.component.html',
+  styleUrls: ['./smb-cluster-list.component.scss']
+})
+export class SmbClusterListComponent extends ListWithDetails implements OnInit {
+  @ViewChild('table', { static: true })
+  table: TableComponent;
+
+  columns: CdTableColumn[];
+  permission: Permission;
+  selection = new CdTableSelection();
+  summaryDataSubscription: Subscription;
+  viewCacheStatus: any;
+  smbClusters: any[];
+  tableActions: CdTableAction[];
+
+  constructor(
+    private authStorageService: AuthStorageService,
+    public actionLabels: ActionLabelsI18n,
+    private smbService: SmbService
+  ) {
+    super();
+
+    this.permission = this.authStorageService.getPermissions().smb;
+    this.tableActions = [
+      {
+        name: this.actionLabels.CREATE + ' Cluster ',
+        permission: 'create',
+        icon: Icons.add,
+        click: () => this.openModal(),
+        canBePrimary: (selection: CdTableSelection) => !selection.hasSingleSelection
+      },
+      {
+        name: this.actionLabels.CREATE + ' Share ',
+        permission: 'create',
+        icon: Icons.add,
+        click: () => this.openModal(),
+        canBePrimary: (selection: CdTableSelection) => !selection.hasSelection
+      },
+      {
+        name: this.actionLabels.EDIT,
+        permission: 'update',
+        icon: Icons.edit,
+        click: () => this.openModal()
+      },
+      {
+        name: this.actionLabels.REMOVE,
+        permission: 'read',
+        icon: Icons.bars,
+        click: () => this.openModal()
+      }
+    ];
+  }
+
+  ngOnInit() {
+    this.columns = [
+      {
+        name: $localize`Cluster`,
+        prop: 'cluster_id',
+        flexGrow: 2
+      },
+      {
+        name: $localize`Auth Mode`,
+        prop: 'auth_mode',
+        flexGrow: 2
+      },
+      {
+        name: $localize`Intent`,
+        prop: 'intent',
+        flexGrow: 2
+      }
+    ];
+  }
+
+  openModal() {
+    throw new Error('Method not implemented.');
+  }
+
+  loadSMBCluster(context: CdTableFetchDataContext) {
+    this.smbService.list().subscribe(
+      (resp: any[]) => {
+        this.smbClusters = resp;
+      },
+      () => {
+        context.error();
+      }
+    );
```

Comment:
I'd encourage you to use more of [async pipes ](https://medium.com/@softwaretechsolution/async-pipe-in-angular-bf0c691faaf2)and less of subscription. You can read [more on that here](https://medium.com/isop-nepal/subscribe-vs-async-pipe-in-angular-21bb38f3ee49). Maybe once we upgrade to newer angular version, we should [switch to using signals](https://angular.dev/guide/signals).


inside ngOnInit() you can use a subject like
```
  
   subject$ = new BehaviorSubject<any>([]);
   smbClusters$: Observable<any>;

    this.smbClusters$ = this.subject$.pipe(
      switchMap(() => this.smbService.list())
    )
```

You can even use a pipe to list() to catch the errors and set the context.

and then on loadSmbCluster() you just call the next of the subject `this.subject$.next([]);`.

after that you can use an async pipe in the html template to fetch the data.

Reply:
Correct.